src/bin/cargo-plumbing/plumbing/resolve_features.rs

```diff
@@ -0,0 +1,149 @@
+    /// Activate all available features
+    #[arg(long)]
+    all_features: bool,
+    /// Do not activate the `default` feature
+    #[arg(long)]
+    no_default_features: bool,
+    /// Include dev units
+    #[arg(long, default_value_t = false)]
+    dev_units: bool,
+    /// Target triple
+    #[arg(long)]
+    target: Vec<String>,
+}
+
+pub(crate) fn exec(gctx: &mut GlobalContext, args: Args) -> CargoResult<()> {
+    let stdin = io::stdin();
+    if stdin.is_terminal() {
+        anyhow::bail!("input must be piped from a file or another command");
+    }
+
+    let manifest_path = args
+        .manifest_path
+        .unwrap_or(env::current_dir()?.join("Cargo.toml"));
+    let path = gctx.cwd().join(manifest_path);
+    let ws = Workspace::new(&path, gctx)?;
+
+    let messages = ResolveFeaturesIn::parse_stream(BufReader::new(stdin));
+
+    let mut lock_version = None;
+    let mut locked_packages = Vec::new();
+    let mut unused_patches = None;
+    let mut specs = Vec::new();
+
+    for message in messages {
+        match message? {
+            ResolveFeaturesIn::Lockfile { version } => lock_version = version,
+            ResolveFeaturesIn::LockedPackage { package } => locked_packages.push(package),
+            ResolveFeaturesIn::UnusedPatches { unused } => unused_patches = Some(unused),
+            ResolveFeaturesIn::Manifest { id } => specs.push(id),
+        }
+    }
+
+    if locked_packages.is_empty() {
+        anyhow::bail!("incomplete input. no packages found.");
+    }
+
+    let resolve = into_resolve(
+        &ws,
+        lock_version,
+        locked_packages,
+        unused_patches.unwrap_or_default(),
+    )?;
+    let cli_features = CliFeatures::from_command_line(
+        &args.features,
+        args.all_features,
+        !args.no_default_features,
+    )?;
+    let has_dev_units = if args.dev_units {
+        HasDevUnits::Yes
+    } else {
+        HasDevUnits::No
+    };
+    let mut registry = ws.package_registry()?;
+    let add_patches = true;
+
+    let resolve_with_overrides = resolve_with_previous(
+        &mut registry,
+        &ws,
+        &cli_features,
+        has_dev_units,
+        Some(&resolve),
+        None,
+        &specs,
+        add_patches,
+    )?;
+
+    let requested_kinds = CompileKind::from_requested_targets(gctx, &args.target)?;
+    let mut target_data = RustcTargetData::new(&ws, &requested_kinds)?;
+    let members_with_features = ws.members_with_features(&specs, &cli_features)?;
+    let member_ids = members_with_features
+        .iter()
+        .map(|(p, _fts)| p.package_id())
+        .collect::<Vec<_>>();
+    let force_all_targets = cargo::core::resolver::ForceAllTargets::No;
+    let pkg_set = get_resolved_packages(&resolve_with_overrides, registry)?;
+    pkg_set.download_accessible(
+        &resolve_with_overrides,
+        &member_ids,
+        has_dev_units,
+        &requested_kinds,
+        &target_data,
+        force_all_targets,
+    )?;
```

Comment:
We should probably have a `// HACK:` comment about it not being intentional that this command does a `cargo fetch` but that its a workaround for https://github.com/rust-lang/cargo/issues/15834